mainloop.go

```diff
@@ -105,53 +97,52 @@ func (m *MainLoop) runMainLoop(ctx context.Context) *govnr.ForeverHandle {
 }
 
 func (m *MainLoop) run(ctx context.Context) {
+	defer m.worker.interrupt()
+
 	if m.electionScheduler == nil {
 		panic("Election trigger was not configured, cannot run Lean Helix (mainloop.run)")
 	}
 
 	m.logger.Info("LHFLOW LHMSG MAINLOOP START LISTENING NOW")
-	workerCtx, cancelWorkerContext := context.WithCancel(ctx)
-	defer cancelWorkerContext()
-	var lastReceivedHeight primitives.BlockHeight
-	for {
+
+	var maxBlockHeightBySync *primitives.BlockHeight
+	var shutdown bool
+	for !shutdown {
+		m.state.GcOldContexts()
 		select {
 		case <-ctx.Done(): // system shutdown
```

Comment:
Not sure I follow what happens on system shutdown. Is there a possible race between old context cancellation a new stuck context ( workerloop running) and system shutdown?

Reply:
GcOldContexts is only for garbage collection of contexts from previous heights. On shutdown there's a call to state.Contexts.Shutdown() which will cancel all worker contexts (present and future).